drivers/sensor/MLX90640.zig

```diff
@@ -0,0 +1,922 @@
+
+        var ta4: f32 = (ta + 273.15);
+        ta4 = ta4 * ta4;
+        ta4 = ta4 * ta4;
+        var tr4: f32 = (tr + 273.15);
+        tr4 = tr4 * tr4;
+        tr4 = tr4 * tr4;
+        const taTr: f32 = tr4 - (tr4 - ta4) / self.emissivity;
+
+        const ktaScale: f32 = @floatFromInt(std.math.pow(u16, 2, self.params.ktaScale));
+        const kvScale: f32 = @floatFromInt(std.math.pow(u16, 2, self.params.kvScale));
+        const alphaScale: f32 = @floatFromInt(std.math.pow(u16, 2, self.params.alphaScale));
+
+        var alphaCorrR: [4]f32 = undefined;
+        alphaCorrR[0] = 1 / (1 + self.params.ksTo[0] * 40);
+        alphaCorrR[1] = 1;
+        var ct: f32 = @floatFromInt(self.params.ct[2]);
+        alphaCorrR[2] = (1 + self.params.ksTo[1] * ct);
+        ct = @floatFromInt(self.params.ct[3] - self.params.ct[2]);
+        alphaCorrR[3] = alphaCorrR[2] * (1 + self.params.ksTo[2] * (ct));
+
+        var gain: f32 = @floatFromInt(self.frame[778]);
+        if (gain > 32767) {
+            gain = gain - 65536;
+        }
+
+        const gee: f32 = @floatFromInt(self.params.gainEE);
+        gain = gee / gain;
+
+        const mode: f32 = @floatFromInt((self.frame[832] & 0x1000) >> 5);
+
+        var irDataCP = [2]f32{
+            @floatFromInt(self.frame[776]),
+            @floatFromInt(self.frame[808]),
+        };
+
+        for (0..2) |i| {
+            if (irDataCP[i] > 32767) {
+                irDataCP[i] = irDataCP[i] - 65536;
+            }
+            irDataCP[i] = irDataCP[i] * gain;
+        }
+
+        var cpo: f32 = @floatFromInt(self.params.cpOffset[0]);
+        irDataCP[0] = irDataCP[0] - cpo * (1 + self.params.cpKta * (ta - 25)) * (1 + self.params.cpKv * (vdd - 3.3));
+
+        const cmee: f32 = @floatFromInt(self.params.calibrationModeEE);
+        cpo = @floatFromInt(self.params.cpOffset[1]);
+        if (mode == cmee) {
+            irDataCP[1] = irDataCP[1] - cpo * (1 + self.params.cpKta * (ta - 25)) * (1 + self.params.cpKv * (vdd - 3.3));
+        } else {
+            irDataCP[1] = irDataCP[1] - (cpo + self.params.ilChessC[0]) * (1 + self.params.cpKta * (ta - 25)) * (1 + self.params.cpKv * (vdd - 3.3));
+        }
+
+        var range: u8 = 0;
+        var pattern: i32 = 0;
+        var pixelNumber: i32 = 0;
+        for (0..768) |i| {
+            pixelNumber = @intCast(i);
+            const ilPattern: i32 = @divTrunc(pixelNumber, 32) - @divTrunc(pixelNumber, 64) * 2;
+            const chessPattern: i32 = ilPattern ^ (pixelNumber - @divTrunc(pixelNumber, 2) * 2);
+            const conversionPattern: i32 = (@divTrunc((pixelNumber + 2), 4) - @divTrunc((pixelNumber + 3), 4) + @divTrunc((pixelNumber + 1), 4) - @divTrunc(pixelNumber, 4)) * (1 - 2 * ilPattern);
+
+            if (mode == 0) {
+                pattern = ilPattern;
+            } else {
+                pattern = chessPattern;
+            }
+
+            var irData: f32 = @floatFromInt(self.frame[i]);
+            if (irData > 32767) {
```

Comment:
Doesn't this just mean that frame should have `i16` instead of `u16`?

Reply:
Yes, it does seem like that.  The conversions that I copied have several instances where this type of logic happens.  I'll try changing to signed ints to see what happens after I get all this refactoring in a good place.

Reply:
This logic, actually, is straight from the datasheet.  Throughout all of the datasheet's calculations there are examples of checking to see if a value is larger than some number then subtracting to make it right.  Due to the extreme complexity of these calculations I think I'd rather leave things as they are.